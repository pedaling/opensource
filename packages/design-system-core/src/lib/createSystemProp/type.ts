import type { Theme } from '../types';

export type InterpolationFn = (props: Record<string, any>) => Record<string, any>;

export type SystemPropConfig = {
  property: string;
  styleProperty?: string;
  scale?: keyof Pick<Theme, 'colors' | 'space' | 'opacity'>;
  disabled?: boolean;
  transform?: (value: any) => Record<string, any> | undefined;
};

export type SystemProp = {
  (props: Record<string, any>): Record<string, any>;
  propName: string;
  disabled: boolean;
};
