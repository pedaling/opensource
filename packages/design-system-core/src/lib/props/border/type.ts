import type { ResponsiveValue } from '../../types';

export type BorderProps = {
  borderWidth?: ResponsiveValue<number>;
  borderStyle?: ResponsiveValue<'solid' | 'none'>;
  borderColor?: ResponsiveValue<string>;
  borderTopWidth?: ResponsiveValue<number>;
  borderTopStyle?: ResponsiveValue<'solid' | 'none'>;
  borderTopColor?: ResponsiveValue<string>;
  borderRightWidth?: ResponsiveValue<number>;
  borderRightStyle?: ResponsiveValue<'solid' | 'none'>;
  borderRightColor?: ResponsiveValue<string>;
  borderBottomWidth?: ResponsiveValue<number>;
  borderBottomStyle?: ResponsiveValue<'solid' | 'none'>;
  borderBottomColor?: ResponsiveValue<string>;
  borderLeftWidth?: ResponsiveValue<number>;
  borderLeftStyle?: ResponsiveValue<'solid' | 'none'>;
  borderLeftColor?: ResponsiveValue<string>;
  borderRadius?: ResponsiveValue<number | string>;
  borderTopLeftRadius?: ResponsiveValue<number | string>;
  borderTopRightRadius?: ResponsiveValue<number | string>;
  borderBottomLeftRadius?: ResponsiveValue<number | string>;
  borderBottomRightRadius?: ResponsiveValue<number | string>;
  outlineWidth?: ResponsiveValue<number>;
  outlineStyle?: ResponsiveValue<'solid' | 'none'>;
  outlineColor?: ResponsiveValue<string>;
  outlineOffset?: ResponsiveValue<number>;
};
