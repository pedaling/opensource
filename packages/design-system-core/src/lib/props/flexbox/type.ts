import type { ResponsiveValue } from '../../types';

export type FlexboxProps = {
  flex?: ResponsiveValue<string | number>;
  flexGrow?: ResponsiveValue<string | number>;
  flexShrink?: ResponsiveValue<string | number>;
  flexBasis?: ResponsiveValue<string | number>;
  flexDirection?: ResponsiveValue<'column' | 'column-reverse' | 'row' | 'row-reverse'>;
  flexWrap?: ResponsiveValue<'nowrap' | 'wrap' | 'wrap-reverse'>;
  alignContent?: ResponsiveValue<
    | 'space-around'
    | 'space-between'
    | 'space-evenly'
    | 'stretch'
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'start'
    | 'baseline'
    | 'normal'
  >;
  alignItems?: ResponsiveValue<
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'self-end'
    | 'self-start'
    | 'start'
    | 'baseline'
    | 'normal'
    | 'stretch'
  >;
  alignSelf?: ResponsiveValue<
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'self-end'
    | 'self-start'
    | 'start'
    | 'auto'
    | 'baseline'
    | 'normal'
    | 'stretch'
  >;
  justifyContent?: ResponsiveValue<
    | 'space-around'
    | 'space-between'
    | 'space-evenly'
    | 'stretch'
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'start'
    | 'left'
    | 'normal'
    | 'right'
  >;
  justifyItems?: ResponsiveValue<
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'self-end'
    | 'self-start'
    | 'start'
    | 'baseline'
    | 'left'
    | 'legacy'
    | 'normal'
    | 'right'
    | 'stretch'
  >;
  justifySelf?: ResponsiveValue<
    | 'center'
    | 'end'
    | 'flex-end'
    | 'flex-start'
    | 'self-end'
    | 'self-start'
    | 'start'
    | 'auto'
    | 'baseline'
    | 'left'
    | 'normal'
    | 'right'
    | 'stretch'
  >;
  order?: ResponsiveValue<number>;
};
