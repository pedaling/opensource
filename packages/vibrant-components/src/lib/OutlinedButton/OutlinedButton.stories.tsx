import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { Icon } from '@vibrant-ui/icons';
import { VStack } from '../VStack';
import { OutlinedButton } from './OutlinedButton';

export default {
  title: 'OutlinedButton',
  component: OutlinedButton,
  args: {
    size: ['sm', 'md', 'md'],
    IconComponent: Icon.Photo.Regular,
    children: 'click me',
  },
} as ComponentMeta<typeof OutlinedButton>;

export const Basic: ComponentStory<typeof OutlinedButton> = props => (
  <VStack m={20}>
    <OutlinedButton {...props} />
  </VStack>
);
