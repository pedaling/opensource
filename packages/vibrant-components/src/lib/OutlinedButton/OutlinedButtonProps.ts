import type { Ref } from 'react';
import type { ResponsiveValue, TextChildren } from '@vibrant-ui/core';
import { propVariant, withVariation } from '@vibrant-ui/core';
import type { IconComponent, IconProps } from '@vibrant-ui/icons';
import { Icon } from '@vibrant-ui/icons';
import type { PressableProps } from '../Pressable';

export type OutlinedButtonProps = {
  ref?: Ref<any>;
  size: ResponsiveValue<'lg' | 'md' | 'sm' | 'xl'>;
  type?: PressableProps['buttonType'];
  IconComponent?: IconComponent<IconProps, 'Fill' | 'Regular'>;
  full?: ResponsiveValue<boolean>;
  active?: boolean;
  disclosure?: boolean;
  disabled?: boolean;
  loading?: boolean;
  onClick?: PressableProps['onClick'];
  children: TextChildren;
  testId?: string;
};

export const withOutlinedButtonVariation = withVariation<OutlinedButtonProps>('OutlinedButton')(
  propVariant({
    props: [
      {
        name: 'size',
        responsive: true,
      },
    ],
    variants: {
      sm: {
        typography: 'body4',
        py: 6,
        px: 7,
        spinnerSize: 'sm',
        iconSize: 16,
        disclosureSize: 16,
        contentsSpacing: 0,
      },
      md: {
        typography: 'body2',
        py: 9,
        px: 9,
        spinnerSize: 'md',
        iconSize: 18,
        disclosureSize: 16,
        contentsSpacing: 0,
      },
      lg: {
        typography: 'body2',
        py: 12,
        px: 11,
        spinnerSize: 'md',
        iconSize: 18,
        disclosureSize: 16,
        contentsSpacing: 0,
      },
      xl: {
        typography: 'body1',
        py: 14,
        px: 13,
        spinnerSize: 'md',
        iconSize: 20,
        disclosureSize: 18,
        contentsSpacing: 2,
      },
    } as const,
  }),
  propVariant({
    props: [
      {
        name: 'disabled',
        keep: true,
        default: false,
      },
    ],
    variants: {
      true: {
        backgroundColor: 'disable',
        borderColor: 'outlineDisable',
      },
      false: {
        backgroundColor: 'surface2',
        borderColor: 'outline1',
      },
    },
  }),
  propVariant({
    props: [
      {
        name: 'full',
        responsive: true,
        default: false,
      },
    ],
    variants: {
      true: {
        width: '100%',
      },
      false: {
        width: 'auto',
      },
    },
  }),
  propVariant({
    props: [
      {
        name: 'disclosure',
      },
      {
        name: 'active',
      },
    ],
    variants: ({ disclosure, active }) => {
      if (!disclosure) {
        return {
          DisclosureIconComponent: null,
        };
      }

      if (active) {
        return { DisclosureIconComponent: Icon.ArrowTriangleUp.Fill };
      }

      return { DisclosureIconComponent: Icon.ArrowTriangleDown.Fill };
    },
  })
);
