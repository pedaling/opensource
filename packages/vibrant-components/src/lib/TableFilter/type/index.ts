export type StringFilterOperator = 'contains' | 'empty' | 'equals' | 'notContains' | 'notEmpty' | 'notEquals';

export type StringFilter = {
  value: string;
  operator: StringFilterOperator[];
};

export type Filter = StringFilter & { dataKey: string };
