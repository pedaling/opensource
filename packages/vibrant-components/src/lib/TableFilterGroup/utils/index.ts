import type { DateFilterOperator, MultiSelectFilterOperator, Option, StringFilterOperator } from '../types';

export const isValueRequiredOperator = (op: string) => op === 'empty' || op === 'notEmpty';

export const isStringFilterValid = (filter: { value: string; operator: StringFilterOperator }) =>
  Boolean(filter.value) || isValueRequiredOperator(filter.operator);

export const isDateFilterValid = (filter: { value: Date[]; operator: DateFilterOperator }) => {
  if (isValueRequiredOperator(filter.operator)) return true;

  if (filter.operator === 'between') return filter.value.length >= 2;

  return filter.value.length >= 1;
};

export const isMultiSelectFilterValid = (filter: { value: Option['value'][]; operator: MultiSelectFilterOperator }) =>
  Boolean(filter.value.length > 0) || isValueRequiredOperator(filter.operator);
