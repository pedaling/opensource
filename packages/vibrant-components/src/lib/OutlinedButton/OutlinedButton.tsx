import { Box, Text } from '@vibrant-ui/core';
import { HStack } from '../HStack';
import { Pressable } from '../Pressable';
import { Spinner } from '../Spinner';
import { withOutlinedButtonVariation } from './OutlinedButtonProps';

export const OutlinedButton = withOutlinedButtonVariation(
  ({
    innerRef,
    type = 'button',
    backgroundColor,
    borderColor,
    IconComponent,
    DisclosureIconComponent,
    loading,
    disabled,
    typography,
    spinnerSize,
    iconSize,
    disclosureSize,
    children,
    contentsSpacing,
    ...restProps
  }) => (
    <Pressable
      {...restProps}
      ref={innerRef}
      as="button"
      buttonType={type}
      interactions={['hover', 'focus', 'active']}
      overlayColor="onView1"
      backgroundColor={backgroundColor}
      borderColor={borderColor}
      borderWidth={1}
      borderRadiusLevel={1}
      disabled={loading || disabled}
    >
      <>
        <HStack
          as="span"
          alignVertical="center"
          alignHorizontal="center"
          opacity={loading ? 0 : 1}
          spacing={contentsSpacing}
        >
          {IconComponent && (
            <Box as="span" mx={2}>
              <IconComponent size={iconSize} />
            </Box>
          )}
          {Boolean(children) && (
            <Text typography={typography} fontWeight="bold" mx={4}>
              {children}
            </Text>
          )}
          {DisclosureIconComponent && (
            <Box as="span" ml={2}>
              <DisclosureIconComponent size={disclosureSize} />
            </Box>
          )}
        </HStack>
        {loading && (
          <HStack
            as="span"
            alignVertical="center"
            alignHorizontal="center"
            position="absolute"
            top={0}
            left={0}
            right={0}
            bottom={0}
          >
            <Spinner size={spinnerSize} />
          </HStack>
        )}
      </>
    </Pressable>
  )
);
