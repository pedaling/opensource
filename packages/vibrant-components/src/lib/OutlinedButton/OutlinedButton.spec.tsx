import type { ReactRenderer } from '@vibrant-ui/utils/testing-web';
import { BrowserTesting, createReactRenderer } from '@vibrant-ui/utils/testing-web';
import { OutlinedButton } from './OutlinedButton';
import type { OutlinedButtonProps } from './OutlinedButtonProps';

describe('<OutlinedButton />', () => {
  const { render } = createReactRenderer();

  let props: OutlinedButtonProps = {
    size: 'md',
    disabled: false,
    full: false,
    disclosure: false,
    children: 'OutlinedButton',
  };

  let subject = () => {
    const renderer = render(<OutlinedButton {...props} />);

    return renderer.container;
  };

  describe('when the size property is responsive', () => {
    subject = () => {
      const renderer = render(<OutlinedButton {...props} />);

      return renderer.getByText(props.children?.toString() ?? '');
    };

    beforeEach(async () => {
      props = {
        ...props,
        size: ['sm', 'md', 'lg'],
      };
    });

    describe('and the viewport is mobile', () => {
      beforeAll(async () => {
        await BrowserTesting.resize(320, 768);
      });

      it('should have bold font weight', async () => {
        expect(subject()).toHaveStyleRule('font-weight', '700');
      });

      it('should not have font weight in media query', async () => {
        expect(subject()).not.toHaveStyleRule('font-weight', expect.any(String), {
          media: '(min-width: 0px)',
        });
      });
    });

    describe('and the viewport is tablet', () => {
      beforeAll(async () => {
        await BrowserTesting.resize(640, 768);
      });

      it('should have bold font weight', async () => {
        expect(subject()).toHaveStyleRule('font-weight', '700');
      });

      it('should not have font weight in media query', async () => {
        expect(subject()).not.toHaveStyleRule('font-weight', expect.any(String), {
          media: '(min-width: 640px)',
        });
      });
    });
  });

  describe('when href provided', () => {
    let renderer: ReactRenderer;

    beforeEach(() => {
      renderer = render(<OutlinedButton {...props} href="https://www.vibrant-design.com" />);
    });

    it('should render link with href attribute', () => {
      expect(renderer.getByRole('link').getAttribute('href')).toBe('https://www.vibrant-design.com');
    });
  });

  it('match snapshot', async () => {
    expect(subject()).toMatchSnapshot();
  });
});
