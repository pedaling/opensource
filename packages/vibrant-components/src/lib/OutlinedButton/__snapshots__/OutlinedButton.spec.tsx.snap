// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OutlinedButton /> match snapshot 1`] = `
.emotion-0 {
  text-align: left;
  color: #0c0c0c;
  -webkit-text-fill-color: #0c0c0c;
  text-fill-color: #0c0c0c;
  font-size: 0.75rem;
  line-height: 1rem;
  font-weight: 700;
  margin-left: 4px;
  margin-right: 4px;
}

@media screen and (min-width: 640px) {
  .emotion-0 {
    font-size: 0.875rem;
    line-height: 1.125rem;
  }
}

<span
  class="emotion-0"
>
  OutlinedButton
</span>
`;
